### PR TITLE
Extract cache protocol.

### LINKED ISSUE #2750: Separate message parsing from interpretation in the cache daemon
So that it's easier to version the protocol

### PR DESCRIPTION
**Pull request by mefyl**
Fix #2750.

### CODE REVIEW

File: src/dune_manager/dune [hunk 1]
```diff
@@ -1,5 +1,5 @@
 (library
  (name dune_manager)
- (modules utils dune_manager)
+ (modules utils dune_manager messages messages_intf)
```
Comment (diml) [inline src/dune_manager/dune]: We can remove this field altogether now

File: src/dune_manager/messages_intf.ml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  ; files : (Path.Build.t * Digest.t) list
+  ; metadata : Sexp.t list
+  ; repository : int option
+  }
+
+type outgoing = Outgoing
+
+type incoming = Incoming
+
+type _ message =
```
Comment (diml) [inline src/dune_manager/messages_intf.ml]: Why use a GADT rather than two separate types?

Comment (mefyl) [reply src/dune_manager/messages_intf.ml]: To have some functions that can work on messages regardless of the direction (e.g. `val send : _ message -> out_channel -> unit`)

File: src/dune_manager/dune [hunk 3]
```diff
@@ -1,5 +1,4 @@
 (library
  (name dune_manager)
- (modules utils dune_manager messages messages_intf)
```
Comment (rgrinberg) [inline src/dune_manager/dune]: I think I already did this change in master btw.

**Pull request merged**